pkg/cli/options.go

```diff
@@ -139,3 +150,150 @@ func WithCompletion() Option {
+func DiscoverExternalPlugins() (ps []plugin.Plugin, err error) {
+	pluginsRoot := setPluginsRootBasedonHost()
+
+	rootInfo, err := os.Stat(pluginsRoot)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			logrus.Debugf("External plugins dir %q does not exist, skipping external plugin parsing", pluginsRoot)
+			return nil, nil
+		}
+		return nil, err
+	}
+	if !rootInfo.IsDir() {
+		logrus.Debugf("External plugins path %q is not a directory, skipping external plugin parsing", pluginsRoot)
+		return nil, nil
+	}
+
+	infos, err := ioutil.ReadDir(pluginsRoot)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, info := range infos {
+		if !info.IsDir() {
+			logrus.Debugf("%q is not a directory so skipping parsing", info.Name())
+			continue
+		}
+
+		subinfos, err := ioutil.ReadDir(filepath.Join(pluginsRoot, info.Name()))
+		if err != nil {
+			return nil, err
+		}
+
+		for _, subinfo := range subinfos {
+			pluginInfos, err := ioutil.ReadDir(filepath.Join(pluginsRoot, info.Name(), subinfo.Name()))
+			if err != nil {
+				return nil, err
+			}
+			for _, pluginInfo := range pluginInfos {
+				// find the executable that matches the same name as info.Name().
+				// if no match is found, compare the external plugin string name before dot and match it with info.Name() which is the external plugin root dir.
+				// for example: sample.sh --> sample, externalplugin.py --> externalplugin
+				trimmedPluginName := strings.Split(pluginInfo.Name(), ".")
+				if len(trimmedPluginName) == 0 {
+					return nil, fmt.Errorf("Invalid plugin name found %q", pluginInfo.Name())
+				}
+
+				if pluginInfo.Name() == info.Name() || trimmedPluginName[0] == info.Name() {
+					// check whether the external plugin is an executable.
+					if !isPluginExectuable(pluginInfo.Mode()) {
+						return nil, fmt.Errorf("External plugin %s found in path is not an executable", pluginInfo.Name())
+					}
+
+					_, args := parseExternalPluginArgs()
+
+					ep := external.Plugin{
+						PName: info.Name(),
+						Path:  filepath.Join(pluginsRoot, info.Name(), subinfo.Name(), pluginInfo.Name()),
+						// todo(rashmigottipati): external.Plugin must satisfy the Plugin interface,
+						// so adding the supported project version as a placeholder value for now.
+						PSupportedProjectVersions: []config.Version{cfgv2.Version, cfgv3.Version},
+						Args:                      args,
+					}
+
+					if err := ep.PVersion.Parse(subinfo.Name()); err != nil {
+						return nil, err
+					}
+
+					logrus.Printf("Adding external plugin: %s", ep.Name())
+
+					ps = append(ps, ep)
+
+				}
+			}
+		}
+
+	}
+	return ps, nil
+}
+
+// isPluginExectuable checks if a plugin is an executable based on the bitmask and returns true or false.
+func isPluginExectuable(mode os.FileMode) bool {
+	return mode&0111 == 0111
+}
+
+// getUserHomeDir returns $XDG_CONFIG_HOME if set, otherwise $HOME.
+func getUserHomeDir() (string, error) {
```

Comment:
nit: I think I would've just used `getHomeDir()`. No need to change this just adding here for posterity.